### ISSUE TITLE
external module argument

### ISSUE THREAD

**Issue by o-sleep** [saltstack/salt#31030]
I am trying to set something up where I trigger an event with the given id as a payload.  Then I want to use this id to retrieve indicative data from an external pillar which calls an external rest interface from a non-saltstack related service.  I just want to be able to pass a value from the event.send payload to the ext_pillar() function so I can modify the rest url.

So far I can see the event payload data in state.event in both the new and ret event.  I have verified that the reactor state file can lookup the payload data by using data.data['mykey'].  I haven't been able to see this value in the ext_pillar() of the external pillar.  I've tried looking at all globals in the function, but don't see it.

I could ditch the external pillar approach and just have another service that pulls the rest interface based on the event, and then kicks off an event or call in saltstack, but was hoping there was a way to do this natively.  I am happy to hear if I am approach this the wrong way.

Thanks,
Bjorn

**Comment by basepi** [comment]
External pillar is not expected to receive any of that data. In fact, it receives very little outside of the arguments with which it is called. You could maybe call the external pillar directly from within a custom state module, which receives the data required and passes it into the external pillar call....

**Comment by o-sleep** [comment]
Ah, so create a module that calls the ext_pillar() function and call it from the reactor state file?

**Comment by basepi** [comment]
Indeed. That might be the direction I'd go if external pillar is an important piece of it. You could also go around the external pillar and just call the pillar backend directly.

**Comment by o-sleep** [comment]
I'm not entirely sure what you mean, unless you are saying to make the call externally to salt and then just feed it events with the indicative data as the payload.  I have thought about doing that and will keep that as a backup.  If you meant something else, please elaborate.  Otherwise, your earlier suggestion gives me a path to work towards.  I'll see how far I get with it and let you know if I get stuck.  Thanks for your help.

**Comment by o-sleep** [comment]
After working on this a bit.  Is there a way I can import the the external pillar to run ext_pillar()?  I tried calling it and mypillar.ext_pillar(), where mypillar is the name of a version of cmd_json, I modified.  But both get "is not defined" errors.  Or did you mean for me to call the external data source directly from the state module, and not use the external pillar system in salt?

**Comment by basepi** [comment]
I feel like you should be able to either import the modules directly, or if that fails, import the loader and call the loader on those modules. Did a direct import not work?

**Comment by o-sleep** [comment]
I went down this route and started worrying about all the dependencies I started to pull in, so I shelved it for now and am trying another tact.

I create a custom execution module to make a urllib2 call to our rest interface.  I am trying to call this from a reactor event that takes the result of this for making a cloud.create call.  I am trying to see how best to do this, though.  I know there is a way to create a variable in a state file (and I am assuming I can do the same for a reactor file).  I try to declare in jinja, via:

`{% set myvar = salt['local.mymodule.get']['myval'] %}`

and it errors with AliasedLoader object has no attribute local.mymodule.get.  It seems to work if I use a builtin execution module, like test.ping.  Is there another way to take the results of an execution module and use it for another (within the reactor)?
